File: algorithms/141.c
#include <stdio.h>
#include <stdbool.h>

struct ListNode
{
    int val;
    struct ListNode *next;
};

bool hasCycle(struct ListNode *head)
{
    if ( head == NULL || head->next == NULL )
        return false;
    struct ListNode *slow = head->next, *fast = slow->next;
    while ( slow != fast && fast && fast->next )
    {
        slow = slow->next;
        fast = fast->next->next;
    }
    return slow == fast;
}

int main()
{
    return 0;
}

File: algorithms/python/141.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        p, q = head, head
        while p and q and q.next:
            p = p.next
            q = q.next.next
            if p == q: return True
        return False
